plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderNodeComparator.java

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.elk.alg.layered.intermediate.preserveorder;
+
+import java.util.Comparator;
+import java.util.List;
+
+import org.eclipse.elk.alg.layered.graph.LEdge;
+import org.eclipse.elk.alg.layered.graph.LNode;
+import org.eclipse.elk.alg.layered.graph.LPort;
+import org.eclipse.elk.alg.layered.graph.Layer;
+import org.eclipse.elk.alg.layered.options.InternalProperties;
+import org.eclipse.elk.alg.layered.options.OrderingStrategy;
+
+/**
+ * Orders {@link LNode}s in the same layer by {@link InternalProperties#MODEL_ORDER}
+ * or the order of the nodes they connect to in the previous layer.
```

Comment:
Is this actually required for layers other than the very first?

Reply:
Yes, the idea is to order the nodes and ports such that they reflect the model order. If I do not do that there might be unnecessary crossings. I do not sort all "free" nodes correctly and use this as an initial configuration for the crossing minimization since this did not seem to be effective.